lint/clang_tidy.bzl

```diff
@@ -0,0 +1,337 @@
+    args.extend(_prefixed(compilation_context.framework_includes.to_list(), "-F"))
+    args.extend(_prefixed(compilation_context.includes.to_list(), "-I"))
+    args.extend(_prefixed(compilation_context.quote_includes.to_list(), "-iquote"))
+    args.extend(_prefixed(compilation_context.system_includes.to_list(), _angle_includes_option(ctx)))
+    args.extend(_prefixed(compilation_context.external_includes.to_list(), "-isystem"))
+
+    return args
+
+def clang_tidy_action(ctx, compilation_context, executable, srcs, stdout, exit_code):
+    """Create a Bazel Action that spawns a clang-tidy process.
+
+    Adapter for wrapping Bazel around
+    https://clang.llvm.org/extra/clang-tidy/
+
+    Args:
+        ctx: an action context OR aspect context
+        compilation_context: from target
+        executable: struct with a clang-tidy field
+        srcs: file objects to lint
+        report: output file containing the stdout or --output-file of clang-tidy
+        exit_code: output file containing the exit code of clang-tidy.
+            If None, then fail the build when clang-tidy exits non-zero.
+    """
+
+    outputs = [stdout]
+    env = {}
+    env["CLANG_TIDY__STDOUT_STDERR_OUTPUT_FILE"] = stdout.path
+    #env["CLANG_TIDY__VERBOSE"] = "1"
+    if exit_code:
+        env["CLANG_TIDY__EXIT_CODE_OUTPUT_FILE"] = exit_code.path
+        outputs.append(exit_code)
+
+    ctx.actions.run_shell(
+        inputs = _gather_inputs(ctx, compilation_context, srcs),
+        outputs = outputs,
+        tools = [executable._clang_tidy_wrapper, executable._clang_tidy],
+        command = executable._clang_tidy_wrapper.path + " $@",
+        arguments = [executable._clang_tidy.path] + _get_args(ctx, compilation_context, srcs),
+        use_default_shell_env = True,
+        env = env,
+        mnemonic = _MNEMONIC,
+        progress_message = "Linting %{label} with clang-tidy",
+    )
+
+def clang_tidy_fix(ctx, compilation_context, executable, srcs, patch, stdout, exit_code):
+    """Create a Bazel Action that spawns clang-tidy with --fix.
+
+    Args:
+        ctx: an action context OR aspect context
+        compilation_context: from target
+        executable: struct with a clang_tidy field
+        srcs: list of file objects to lint
+        patch: output file containing the applied fixes that can be applied with the patch(1) command.
+        stdout: output file containing the stdout or --output-file of clang-tidy
+        exit_code: output file containing the exit code of clang-tidy
+    """
+    patch_cfg = ctx.actions.declare_file("_{}.patch_cfg".format(ctx.label.name))
+
+    args = _get_args(ctx, compilation_context, srcs)
+    ctx.actions.write(
+        output = patch_cfg,
+        content = json.encode({
+            "linter": executable._clang_tidy_wrapper.path,
+            "args": [executable._clang_tidy.path, "--fix"] + _get_args(ctx, compilation_context, srcs),
+            "env": {
+                #"CLANG_TIDY__VERBOSE": "1",
```

Comment:
comment about why this is commented out?

Reply:
It's for debugging. shall I pass a verbose flag through from aspect factory to control this, to make it easier for integrators to see the clang-tidy command lines being invoked?

Reply:
Added this as user configuration via factory option